staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -410,13 +410,50 @@ func (qs *queueSet) syncTimeLocked() {
 	timeSinceLast := realNow.Sub(qs.lastRealTime)
 	qs.lastRealTime = realNow
 	prevR := qs.currentR
-	qs.currentR += SeatsTimesDuration(qs.getVirtualTimeRatioLocked(), timeSinceLast)
-	if qs.currentR < prevR {
-		klog.ErrorS(errors.New("progress meter wrapped around"), "Wrap", "QS", qs.qCfg.Name, "prevR", prevR, "currentR", qs.currentR)
+	incrR := SeatsTimesDuration(qs.getVirtualTimeRatioLocked(), timeSinceLast)
+	if incrR > RDecrement/2 {
+		klog.ErrorS(errors.New("unusably large R increment"), "Can not do it", "QS", qs.qCfg.Name, "when", realNow.Format(nsTimeFmt), "incrR", incrR)
+	}
+	qs.currentR = prevR + incrR
+	switch {
+	case prevR > qs.currentR:
+		klog.ErrorS(errors.New("progress meter overflow"), "Wrap", "QS", qs.qCfg.Name, "when", realNow.Format(nsTimeFmt), "prevR", prevR, "currentR", qs.currentR)
+	case qs.currentR >= HighR:
+		qs.advanceEpoch()
 	}
 	metrics.SetCurrentR(qs.qCfg.Name, qs.currentR.ToFloat())
 }
 
+// advanceEpoch subtracts RDecrement from the global progress meter R
+// and all the readings that have been taked from that meter.
+func (qs *queueSet) advanceEpoch() {
```

Comment:
As an admin - I would really like to be able to monitor this.

I think we need a metric that shows
a) when advancing epoch happened
b) any errors asociated with that (in particular I'm seeing the underflow error below).

Reply:
I am most concerned by the fact that you are reporting seeing an underflow.  That is far beyond what I expect.  Can you give more details and/or describe how to reproduce?

Reply:
It seems I wasn't clear - I didn't see underflow myself. I see that as error below.
But I would like to be able to monitor for that (i.e. have metrics - not just logs).